### PR TITLE
Speed up MKLDNN conv & linear with unary or binary post op by weight cache

### PR DESCRIPTION
**Pull request by Xia-Weiwen**
Speed up MKLDNN conv & linear with unary or binary post op by weight cache. Conv and linear have their own cache. The caches are global.

### CODE REVIEW

**Comment by Xia-Weiwen** [comment]
@Guobing-Chen Please also review. Thanks.

File: aten/src/ATen/native/mkldnn/Conv.cpp [hunk 1]
```diff
@@ -215,6 +315,27 @@ Tensor _mkldnn_convolution(
...
+                 reinterpret_cast<int32_t*>(w_buf_ptr)[weight_t.numel()-1] :
+                 (int32_t)reinterpret_cast<int16_t*>(w_buf_ptr)[weight_t.numel()-1];
+  std::vector<int32_t> weight_data_samples = {sample1, sample2};
+  auto key = create_weight_cache_key(
+      x.get_dims(), w.get_dims(), w_buf_ptr, weight_data_samples,
+      strides, dilates, paddings, paddings, groups, op_attr, memory_format,
+      input_t.scalar_type(), weight_t.scalar_type(), at::get_num_threads());
+  const ideep::tensor& expected_w =
+      get_expected_weight(key, x, w, strides, paddings, paddings,
+                          dilates, groups, op_attr, is_channels_last);
```
Comment (jgong5) [inline aten/src/ATen/native/mkldnn/Conv.cpp:337]: Would this cause problem if the weight is changed (not constant)? What is the assumption you make about it here?

Comment (jgong5) [reply aten/src/ATen/native/mkldnn/Conv.cpp:337]: This is not the right thing to do - either check weight content or check key only. We don't have a reliable way of knowing if weight is constant without the caller telling us.

Comment (Xia-Weiwen) [reply aten/src/ATen/native/mkldnn/Conv.cpp:337]: We have made weight info part of the key. We find packed weight in cache by key. The key is an `std::string` consisting of various info, including input/weight shape, weight buffer address, weight data samples, number of threads, etc. If weight is changed, the key is also changed.

Comment (jgong5) [reply aten/src/ATen/native/mkldnn/Conv.cpp:337]: Well, you know that's not enough, right? Users can change the weight content from the frontend without changing the key you define here.

Comment (Xia-Weiwen) [reply aten/src/ATen/native/mkldnn/Conv.cpp:337]: OK, I will check all elements of the weight tensor. But I think the overhead may be too big to create the key. I can implement it this way first and collect some perf data to validate.

Comment (jgong5) [reply aten/src/ATen/native/mkldnn/Conv.cpp:337]: No, I don't think scanning weights is the right thing to do.

**Comment by Xia-Weiwen** [comment]
We will implement this in the frontend instead of backend (as in this PR). Will file new PR to do it.

**Pull request closed without merging**